### ISSUE TITLE
End game button...?

### ISSUE THREAD

**Issue by itshaydendev** [Lattyware/massivedecks#75]
If there is one, I can't find it. I mean END game not LEAVE game.

**Comment by Lattyware** [comment]
No, this is simply not currently there. This is related to #13 which is about automatically ending the game when someone reaches a certain score.

For now, you can simply look at the scores and start a new lobby if you want to go again. This is definitely a feature that would be nice to have.

**Comment by haydenbjyoung** [comment]
Right. Might check out the code and have a crack at making that score feature for ya +1

**Comment by Lattyware** [comment]
I'd be happy to accept pull requests. Feel free to ask about things if you need to.

**Comment by haydenbjyoung** [comment]
I'm gonna set up a dev version over on my VPS. Any way to change the server ports so it doesn't run on :80? I have my website there, so...

**Comment by Lattyware** [comment]
Yes, the play framework supports doing that, see https://stackoverflow.com/a/8206747/722121 for the different ways of doing it. (MD uses play 2.x).

**Comment by Lattyware** [comment]
To be clear, if you are using the gulp script, it is [where `activator` is being run](https://github.com/Lattyware/massivedecks/blob/9b74f7ed4adc49322452945159a89f01ad434c22/gulpfile.js#L42), adding the port to the argument list as suggested in the SO question.

**Comment by haydenbjyoung** [comment]
Okay, thanks :)

**Comment by haydenbjyoung** [comment]
I'll just leave a side note that this will be my first encounter with Scala but I reckon I'll get it.

**Comment by Lattyware** [comment]
It's a nice language - always nice to have an excuse to try something new out. I imagine the server side changes should be relatively minimal for this. You'll probably want to start in [the API controller](https://github.com/Lattyware/massivedecks/blob/master/app/massivedecks/controllers/API.scala) and work from there.

**Comment by haydenbjyoung** [comment]
Alright, will do. I've been browsing code and as it's mainly java-central, it shouldn't be difficult to pick up, so that's good.

**Comment by haydenbjyoung** [comment]
Sorry to disturb, but I'm trying to run `gulp` on the folder, with SBT and Elm installed and after running `npm i`, and it does this: 
```shell
root@haydenbjyoung:/mnt/massivedecks-01# node_modules/.bin/gulp make
[01:23:05] Using gulpfile /mnt/massivedecks-01/gulpfile.js
[01:23:05] Starting 'make:elm'...
[01:23:05] Starting 'make:scala'...
[01:23:05] Finished 'make:scala' after 6.89 ms
events.js:160
      throw er; // Unhandled 'error' event
      ^

Error: spawn activator ENOENT
    at exports._errnoException (util.js:1020:11)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:197:32)
    at onErrorNT (internal/child_process.js:376:16)
    at _combinedTickCallback (internal/process/next_tick.js:80:11)
    at process._tickCallback (internal/process/next_tick.js:104:9)
    at Module.runMain (module.js:606:11)
    at run (bootstrap_node.js:383:7)
    at startup (bootstrap_node.js:149:9)
    at bootstrap_node.js:496:3```

**Comment by Lattyware** [comment]
Looks like it's failing to start `activator` - have you got [the play framework](https://www.playframework.com/documentation/2.5.x/Installing)? MD is still on an older version (2.5.4, looking at [`plugins.sbt`](https://github.com/Lattyware/massivedecks/blob/master/project/plugins.sbt#L2) `activator` has been replaced with just using `sbt` in newer ones), so you might also just have too new a version.

The plan is to update to current versions, currently we are waiting on the new elm version to bump everything at once.

**Comment by haydenbjyoung** [comment]
It was actually down to me not having set the environment variables up right. I forgot to add /opt/sbt/bin and /opt/activator to my PATH variable. It's fixed now.

**Comment by haydenbjyoung** [comment]
Got my server up and running, it's over at http://haydenbjyoung.me:8080

**Comment by haydenbjyoung** [comment]
@Lattyware Right, in `massivedecks.models.Game`, what is the `playingGameStateFormat` value? Because there's a `configuring`, `started` and `finished` case setup in there, and it looks like I could use it if that's just a game state system.

**Comment by Lattyware** [comment]
@haydenbjyoung That looks like the right way to store the information the game has ended. The `gameStateWrites` value is a Formatter for producing JSON from the state object - not something you should need to change unless you change the data structure. `Json.toJson()` will use the required formatters based on the type of whatever you give it.

At the moment, you can see the state is [generated in Lobby](https://github.com/Lattyware/massivedecks/blob/master/app/massivedecks/lobby/Lobby.scala#L65) - and we don't have a case that generates the `Finished` state. You will need to add a case that does so and make sure that when that when the game is finished [a notification](https://github.com/Lattyware/massivedecks/blob/master/app/massivedecks/notifications/Notifiers.scala#L135) is sent to connected clients.
